### ISSUE TITLE
Media button mapping on Android

### ISSUE THREAD

**Issue by h45rd** [xbmc/xbmc#15003]
Googling for the problem in the title after switching from 17 to 18RC2 lead me [this trac issue](https://trac.kodi.tv/ticket/17979#modify) by mong00se.
As there was no reply since August and I am not sure if trac issues are still being worked on, I am reposting it here.
Can confirm everything the original poster said for 18RC2 and also the latest nightly.

> Platform: Android 8.x (Shield Experience 7.x)
> Device: Nvidia Shield TV 2015
> Arch: arm64-v8a
> Kodi Version: tested on all tagged alphas up to 3 and yesterdays nightly
> 
> Switched from 17 to 18a3 and my custom keymappings for the fast-forward and rewind buttons on my remote (MCE compatible) are being ignored.
> They default back to their original behavior. All other remappings work as expected as far as I can tell.
> Key presses on these two buttons also do not produce events in debug logging anymore.
> 
> Debug log snippet from 17 on keypress:
> ```
> DEBUG: Keyboard: scancode: 0x5a, sym: 0x00bb, unicode: 0x0000, modifier: 0x0
> DEBUG: OnKey: fastforward (0xf0c5) pressed, action is Playlist
> DEBUG: Keyboard: scancode: 0x59, sym: 0x00ba, unicode: 0x0000, modifier: 0x0
> DEBUG: OnKey: rewind (0xf0c4) pressed, action is FullScreen
> Debug log snippet from 18 on keypress:
> 
> ```
> 
> Debug log snippet from 18 on keypress:
> ```
> ```

So is this regression considered a fixable bug or the new norm?
Just wondering if it makes sense to wait for a fix or try working around it by creating new keymaps for my devices.

Best,
h45rd

**Comment by da-anda** [comment]
maybe @garbear (since he reworked input handling for v18) or @peak3d might have an idea. Mind having a look guys if you have a spare minute? Thanks.

**Comment by mazeyx** [comment]
i had a key set for stop playback as my stop button doesnt work that well. and this has stopped working also but my other mappings work fine like set watched context etc. on v17.6 it worked fine but not v18.

**Comment by h45rd** [comment]
Thank you for the clarification as to why the media button handling was changed, da-anda.

For me being able to customize all the keys on my remote is preferential to the rather obscure use case of having KODI play music in the background and wanting to stop playback without switching back to KODI first.
But I guess it also depends greatly on the environment and device one uses the application in/with.

Your suggested solution would be great, as it has the potential to make everyone happy :)

**Comment by puppinoo** [comment]
I just installed Leia RC3 on one of my FTV (a Pendant) since my Shield is broken right now and I can confirm fastforward and rewind on my custom keyboard.xml are ignored and default events are triggered instead. I really hope this issue can be fixed.

Thanks.

**Comment by garbear** [comment]
I responded in the forum post.

**Comment by grimpy** [comment]
@garbear I replied there not sure you rather continue here or there...

**Comment by puppinoo** [comment]
Not sure it's the same problem. In the case of Amazon devices the event is basically overridded by FireOS cause it takes control of MediaSession layer. Koying explains this here:
[https://forum.kodi.tv/showthread.php?tid=331475&pid=2732740#pid2732740](url)
It's a very nasty regression IMO.

**Comment by MartijnKaijser** [comment]
Same https://github.com/xbmc/xbmc/issues/15189

**Comment by h45rd** [comment]
Congrats on a great release :)

But since this regression made it into the final I wonder..
Any chances of it still being fixed / changed in the future or is it considered _expected behavior_ now and this issue should be closed?

**Comment by puppinoo** [comment]
> Congrats on a great release :)
> 
> But since this regression made it into the final I wonder..
> Any chances of it still being fixed / changed in the future or is it considered _expected behavior_ now and this issue should be closed?

I hardly believe that losing button functionalities after they worked perfectly can be considered "the new norm". I hope it wont happen like that.

**Comment by h45rd** [comment]
I agree. But I also understand that we might be fringe cases, given that barely anyone else has complained about this.
Let me elaborate on why I am so annoyingly insistent in at least wanting to know what's what and mean no disrespect in any way :)

Over the years I deployed Kodi for a lot of friends on a great variety of devices.
They have one thing in common: my initial configuration and a custom keymap.
While I personally run bleeding edge software on non-critical infrastructure for the fun of it and am used to _breaking changes_, all these "remote managed" devices are on the stable branch.

Long story short.. those boxes will soon be hit with addon incompatibilities and all sorts of other issues that arise when using outdated software. So I have to update them rather sooner than later.
Since that update is going to break parts of my custom configuration and therefore the daily workflow people are accustomed to, I would like to know if there is sense in waiting for a possible fix to this regression.
Or if I should rather just suck it up, prepare to face the music (we all know how much fun it is to explain to not so tech-savvy, maybe even elderly people why something works different now..) and deploy a new keymap while updating to 18 to get it over with :)

**Comment by puppinoo** [comment]
I agree with the fact things can break. I'm an hobbist java developer so... broken things are the peak of my daily routine :D. Broken things can be fixed though. Here we are talking about the chance to make the FireOS minimalist but perfect to me remote completely useless cause this regression won't be fixed. I think it's different. If there are options I gladly accept that, even if there's work to do. If the remote is doomed cause people like to listen to music with Kodi put in BG I will accept that too but sadly.

**Comment by briantho** [comment]
Thought i was one of the few to have this issue. I am running kodi on a fire stick, and primarily use kodi for music playback. It works great, except now in 18 i can't use my custom keymap to map the "fast forward" and "rewind" buttons to skip and replay. I would really love to see this issue solved, and not have it as the "new norm". On 17.6 my custom keymap worked perfect. Now it does not. I was able to create a workaround, until this is fixed without having to revert to 17, by using a media key remapper on the fire tv, but to be honest it is less than ideal. If anyone is interested my workaround, read about it in [this thread. ](https://forum.kodi.tv/showthread.php?tid=339375)

**Comment by h45rd** [comment]
Using a button remapper app (or if you have root access and are comfortable with the shell: custom keychars/keylayouts specifically for the vendor id of your remote control) might be a feasible workaround for those who use their device exclusively for Kodi.

But since this method works on a system-wide basis, it would prevent you from using the key's original function in other apps that do not support custom keymaps. Which includes basically all commercial android apps, e.g. Netflix, Prime, etc.
So depending on your device, use case and remapped buttons the workaround has the potential to break a lot more than it fixes.

The only "low impact" solution that comes to mind would be a separate remote control just for Kodi.. but I would much prefer a revert (or advancedsettings option) to the media button handling of Kodi 17.6.
Either way an official statement on this topic would be much appreciated.

**Comment by basrieter** [comment]
I would agree with @h45rd. Is there anyway to configure this behaviour?

**Comment by h45rd** [comment]
Thanks for working on this @peak3d. Much appreciated :)
Is this the right place to discuss or should I open a new issue/follow up in the PR?

Only did a short test with the latest nightly on my shield and wanted to give some quick feedback/observations:

- Media keys previous default functions have to be manually mapped now (might be a regression for users who never tinkered with custom keymaps)
- There is some weirdness with double taps being interpreted as long button presses; only applies to media keys. Might be exclusive to my setup/rc. Will investigate further once I get some time.

Edit: Took an early break for some tinkering :). Not able to map my **play** media key. Same applies to the stop key.
#### Short press of the play button:
`2019-03-05 10:42:58.816 T:27531   DEBUG: Keyboard: scancode: 0x7e, sym: 0x00b3, unicode: 0x0000, modifier: 0x0`
[not registered by keymap editor plugin]

#### Second press of the play button (interpreted as long press, even if seconds apart):
```
2019-03-05 10:42:58.816 T:27531   DEBUG: Keyboard: scancode: 0x7e, sym: 0x00b3, unicode: 0x0000, modifier: 0x0
2019-03-05 10:43:08.292 T:27531   DEBUG: Keyboard: scancode: 0x7e, sym: 0x00b3, unicode: 0x0000, modifier: 0x0
2019-03-05 10:43:08.292 T:27531   DEBUG: HandleKey: long-play_pause (0x100f0bd) pressed, action is Info
```
[interpreted by keymap editor plugin as key id 16838845]
#### Actual long press of the play button:
Does not generate a debug entry but opens a Google Assistant overlay. No idea where that comes from.

#### Relevant keymap.xml excerpt:
```
   <key id="16838845">PlayPause</key>
   <play_pause>PlayPause</play_pause>
```

As you mentioned FF in the PR: FF/RW seem to behave normally.


On a sidenote: is there a way I can convert the raw debug scancode output to a Kodi key id myself without keymap editor's KeyListener? A quick search only revealed similar unanswered questions.

**Comment by peak3d** [comment]
I'll do a rework of this today. If the PR is ready, it will mentioned here. please try immediately if the PR builds are done, thx.

**Comment by h45rd** [comment]
Quick follow-up as I have to get back to work.. :)
While finished up I noticed that presses of the affected media keys generate differing scancodes, depending on if video is currently playing.

#### Pressing Play twice some seconds apart without video running
```
2019-03-05 11:14:06.399 T:28401   DEBUG: Keyboard: scancode: 0x55, sym: 0x00b3, unicode: 0x0000, modifier: 0x0
2019-03-05 11:14:10.741 T:28401   DEBUG: HandleKey: play_pause (0xf0bd) pressed, action is PlayPause
2019-03-05 11:14:13.313 T:28401   DEBUG: Keyboard: scancode: 0x55, sym: 0x00b3, unicode: 0x0000, modifier: 0x0
2019-03-05 11:14:13.445 T:28401   DEBUG: HandleKey: play_pause (0xf0bd) pressed, action is PlayPause
```
[expected behavior and correct mapping]

#### Pressing Play twice some seconds apart with video running
```
2019-03-05 11:14:38.196 T:28401   DEBUG: Keyboard: scancode: 0x7f, sym: 0x00b3, unicode: 0x0000, modifier: 0x0
2019-03-05 11:14:41.735 T:28401   DEBUG: Keyboard: scancode: 0x7f, sym: 0x00b3, unicode: 0x0000, modifier: 0x0
2019-03-05 11:14:41.736 T:28401   DEBUG: HandleKey: long-play_pause (0x100f0bd) pressed, action is Info
```
[unexpected behavior and different scancode]

**Comment by basrieter** [comment]
I had the exact same issue. I need to double click play/pause to actually play/pause.

**Comment by peak3d** [comment]
Pls. grep the apk form the PR (you find it by clicking on the jenkins build and after that selecting your android platform) and report, if this fixes your current issues, thx!

**Comment by h45rd** [comment]
Did only a quick test, but everything seems to work as expected with #15675.
Thank you very much :)

One oddity remains (not relevant to my use case, mentioning just in case this is not by design):
While in full screen video long press play_pause does not get passed to Kodi (no debug log entries), but Android instead. Which in turn opens a Google Assistant overlay (at least on the Shield).

If no video is running, long press on play_pause works as expected:
```
2019-03-05 18:12:59.271 T:3920   DEBUG: Keyboard: scancode: 0x55, sym: 0x00b3, unicode: 0x0000, modifier: 0x0
2019-03-05 18:12:59.777 T:3920   DEBUG: HandleKey: long-play_pause (0x100f0bd) pressed, action is Info
```

**Comment by basrieter** [comment]
I checked and my problem with the double press for Play/Pause is now solved with  #15675.
